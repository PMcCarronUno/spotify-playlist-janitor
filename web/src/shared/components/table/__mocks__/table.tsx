import React from "react";

export const Table = () => (
  <table>
    <thead>
      <tr>
        <th>Column A</th>
        <th>Column B</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>Row 1 A</td>
        <td>Row 1 B</td>
      </tr>
      <tr>
        <td>Row 2 A</td>
        <td>Row 2 B</td>
      </tr>
      <tr>
        <td>Row 3 A</td>
        <td>Row 3 B</td>
      </tr>
    </tbody>
  </table>
);
