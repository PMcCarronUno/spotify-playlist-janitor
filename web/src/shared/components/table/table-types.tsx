import { ReactNode } from "react";

export type SortOrder = "asc" | "desc";

export type Column = {
  label?: string | ReactNode,
  accessor: string, 
  sortable?: boolean, 
  sortbyOrder?: SortOrder,
  primary?: boolean;
  render?(arg: any): string | ReactNode,
}