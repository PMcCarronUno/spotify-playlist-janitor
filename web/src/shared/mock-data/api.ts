import { LoginResponse } from "api/auth-api/auth-api-types";
import { DatabasePlaylistResponse, SkippedTrackResponse } from "api/data-api";
import { SpotifyPlaylistResponse } from "api/spotify-api";

type AuthData = {
  login: LoginResponse;
};

export const auth: AuthData = {
  login: {
    refresh_token: "access_token",
    token_type: "Bearer",
    access_token: "access_token",
    expires_in: 4503,
  },
};

export const databasePlaylists: DatabasePlaylistResponse[] = [
  {
    id: "1",
    skipThreshold: 10,
    ignoreInitialSkips: false,
    autoCleanupLimit: 10,
  },
  {
    id: "2",
    skipThreshold: 2,
    ignoreInitialSkips: false,
    autoCleanupLimit: 2,
  },
  {
    id: "3",
    skipThreshold: 3,
    ignoreInitialSkips: true,
    autoCleanupLimit: 3,
  },
  {
    id: "4",
    skipThreshold: 4,
    ignoreInitialSkips: true,
    autoCleanupLimit: 4,
  },
];

export const spotifyPlaylists: SpotifyPlaylistResponse[] = [
  {
    id: "1",
    name: "Playlist 1",
    href: "https://open.spotify.com/playlist/1",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/101",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/102",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/103",
      },
    ],
  },
  {
    id: "2",
    name: "Playlist 2",
    href: "https://open.spotify.com/playlist/2",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/201",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/202",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/203",
      },
    ],
  },
  {
    id: "3",
    name: "Playlist 3",
    href: "https://open.spotify.com/playlist/3",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/301",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/302",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/303",
      },
    ],
  },
  {
    id: "4",
    name: "Playlist 4",
    href: "https://open.spotify.com/playlist/4",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/401",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/402",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/403",
      },
    ],
  },
  {
    id: "5",
    name: "Playlist 5",
    href: "https://open.spotify.com/playlist/5",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/501",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/502",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/503",
      },
    ],
  },
  {
    id: "6",
    name: "Playlist 6",
    href: "https://open.spotify.com/playlist/6",
    images: [
      {
        height: 640,
        width: 640,
        url: "https://mosaic.scdn.co/601",
      },
      {
        height: 300,
        width: 300,
        url: "https://mosaic.scdn.co/602",
      },
      {
        height: 60,
        width: 60,
        url: "https://mosaic.scdn.co/603",
      },
    ],
  },
];

export const skippedTracks: SkippedTrackResponse = [
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 01",
    name: "Name 1",
    skippedDate: new Date("2023-10-31T19:51:25Z"),
    artists: [
      {
        id: "artist 01",
        name: "Artist 1",
        href: "https://open.spotify.com/artist/1",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 1",
      href: "https://open.spotify.com/album/1",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/101",
        }
      ]
    }
  },
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 01",
    name: "Name 1",
    skippedDate: new Date("2023-10-31T19:51:38Z"),
    artists: [
      {
        id: "artist 01",
        name: "Artist 1",
        href: "https://open.spotify.com/artist/1",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 1",
      href: "https://open.spotify.com/album/1",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/101",
        }
      ]
    }
  },
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 01",
    name: "Name 1",
    skippedDate: new Date("2023-10-31T19:52:07Z"),
    artists: [
      {
        id: "artist 01",
        name: "Artist 1",
        href: "https://open.spotify.com/artist/1",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 1",
      href: "https://open.spotify.com/album/1",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/101",
        }
      ]
    }
  },
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 02",
    name: "Name 2",
    skippedDate: new Date("2023-10-31T19:53:10Z"),
    artists: [
      {
        id: "artist 02",
        name: "Artist 2",
        href: "https://open.spotify.com/artist/2",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 2",
      href: "https://open.spotify.com/album/2",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/201",
        }
      ]
    }
  },
  {
    playlistId: databasePlaylists[0].id,
    trackId: "track 02",
    name: "Name 2",
    skippedDate: new Date("2023-10-31T19:53:30Z"),
    artists: [
      {
        id: "artist 02",
        name: "Artist 2",
        href: "https://open.spotify.com/artist/2",
      }
    ],
    album: {
      id: "album 01",
      name: "Album 2",
      href: "https://open.spotify.com/album/2",
      images: [
        {
          height: 240,
          width: 240,
          url: "https://mosaic.scdn.co/201",
        }
      ]
    }
  }
];
