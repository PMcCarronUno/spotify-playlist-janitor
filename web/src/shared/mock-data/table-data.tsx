import { Column } from "../components/table";
import { PiClock } from "react-icons/pi";
import { VscTrash } from "react-icons/vsc";

export const tableColumns: Column[] = [
  {
    label: "",
    accessor: "image",
    sortable: false,
    primary: true,
    render: (url: string) => <img alt="" width={40} height={40} src={url} />,
  },
  {
    label: "Title",
    accessor: "title",
    sortable: true,
    sortbyOrder: "desc",
    primary: true,
  },
  { label: "Artist", accessor: "artist", sortable: true },
  { label: "Album", accessor: "album", sortable: true },
  {
    label: <PiClock />,
    accessor: "length",
    sortable: true,
    render: (millis: number) => {
      var minutes = Math.floor(millis / 60000);
      var seconds = parseInt(((millis % 60000) / 1000).toFixed(0));
      return seconds === 60
        ? minutes + 1 + ":00"
        : minutes + ":" + (seconds < 10 ? "0" : "") + seconds;
    },
  },
  {
    label: "Date Skipped",
    accessor: "date_skipped",
    sortable: true,
    render: (date: Date) => date.toISOString(),
  },
  {
    label: "",
    accessor: "delete",
    sortable: false,
    render: (id: string) => (
      <VscTrash
        style={{ cursor: "pointer" }}
        onClick={() => alert(`Deleted track id: ${id}`)}
      />
    ),
  },
];

export const tableData = [
  {
    id: 1,
    image: "https://i.scdn.co/image/ab67616d00004851835a9e77dae1c928f871ac73",
    title: "Longshot",
    artist: "Catfish and the Bottlemen",
    album: "The Balance",
    length: 232960,
    date_skipped: new Date(2023, 9, 15, 15, 30, 45, 20),
    delete: undefined,
  },
  {
    id: 2,
    image: "https://i.scdn.co/image/ab67616d0000485158406b3f1ac3ceaff7a64fef",
    title: "Dark Necessities",
    artist: "Red Hot Chili Peppers",
    album: "The Getaway",
    length: 302000,
    date_skipped: new Date(2023, 9, 15, 15, 30, 46, 25),
    delete: undefined,
  },
  {
    id: 3,
    image: "https://i.scdn.co/image/ab67616d00004851d86a1e021e7acc7c07c7d668",
    title: "Live Forever",
    artist: "Oasis",
    album: "Definitely Maybe",
    length: 302000,
    date_skipped: new Date(2023, 9, 15, 15, 30, 47, 30),
    delete: undefined,
  },
  {
    id: 4,
    image: "https://i.scdn.co/image/ab67616d00004851176ca6e2a6efb0e3aaf4e37a",
    title: "Touch It / Technologic",
    artist: "Daft Punk",
    album: "Alive 2007",
    length: 276666,
    date_skipped: new Date(2023, 9, 15, 15, 30, 47, 40),
    delete: undefined,
  },
];
