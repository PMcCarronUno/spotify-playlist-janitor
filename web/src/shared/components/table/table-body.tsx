import { SecondaryText, Text } from "../typography";
import { Column } from "./table-types";

export const TableBody = ({
  tableData,
  columns,
}: {
  tableData: any[];
  columns: Column[];
}) => {
  return (
    <tbody>
      {tableData.map((data: any, index: number) => {
        return (
          <tr key={data.id}>
            {columns.map(({ accessor, primary, render }: Column) => {
              const tData = data[accessor] ? data[accessor] : "";
              const tRender = render ? render(tData) : tData;
              return (
                <td
                  key={accessor}
                  id={`table-body-cell-${index}-${accessor}`}
                  data-testid={`table-body-cell-${index}-${accessor}`}
                >
                  {primary ? (
                    <Text>{tRender}</Text>
                  ) : (
                    <SecondaryText>{tRender}</SecondaryText>
                  )}
                </td>
              );
            })}
          </tr>
        );
      })}
    </tbody>
  );
};

TableBody.displayName = "TableBody";
