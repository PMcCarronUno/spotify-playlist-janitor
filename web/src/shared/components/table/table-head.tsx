import { useState } from "react";
import styled from "styled-components";
import { VscArrowUp, VscArrowDown } from "react-icons/vsc";

import { TABLE_HOVER } from "shared/constants";

import { Column, SortOrder } from "./table-types";
import { Text } from "../typography";

type TableHeadProps = {
  columns: Column[];
  handleSorting(accessor: string, sortOrder: SortOrder): void;
};

const Tr = styled.tr`
  .sortable {
    cursor: pointer !important;
  }

  .sortable:hover {
    background-color: ${TABLE_HOVER};
  }

  .content {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
  }

  svg {
    padding-top: 3px;
    margin-bottom: -3px;
  }
`;

export const TableHead = ({ columns, handleSorting }: TableHeadProps) => {
  const defaultSortField = columns.find((column) => column.sortbyOrder);
  const [sortField, setSortField] = useState(defaultSortField?.accessor ?? "");
  const [order, setOrder] = useState<SortOrder>(
    defaultSortField?.sortbyOrder ?? "asc"
  );

  const handleSortingChange = (accessor: any) => {
    const sortOrder =
      accessor === sortField && order === "asc" ? "desc" : "asc";
    setSortField(accessor);
    setOrder(sortOrder);
    handleSorting(accessor, sortOrder);
  };

  return (
    <thead>
      <Tr>
        {columns.map(({ label, accessor, sortable }: any) => {
          let icon = undefined;

          if (accessor === sortField) {
            icon = order === "asc" ? <VscArrowUp /> : <VscArrowDown />;
          }
          return (
            <th
              id={`table-head-cell-${accessor}`}
              data-testid={`table-head-cell-${accessor}`}
              key={accessor}
              onClick={
                sortable ? () => handleSortingChange(accessor) : () => {}
              }
              className={sortable ? "sortable" : ""}
            >
              <div className="content">
                <Text>{label}</Text>
                {icon}
              </div>
            </th>
          );
        })}
      </Tr>
    </thead>
  );
};

TableHead.displayName = "TableHead";
