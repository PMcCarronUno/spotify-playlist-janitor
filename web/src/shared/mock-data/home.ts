import { Playlist } from "modules/home/home-types";
import { databasePlaylists, spotifyPlaylists } from "shared/mock-data/api";

export const monitoredPlaylists: Playlist[] = spotifyPlaylists
  .filter((spotifyPlaylist) =>
    databasePlaylists.some(
      (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
    )
  )
  .map((monitoredPlaylist) => ({
    id: monitoredPlaylist.id,
    href: monitoredPlaylist.href,
    name: monitoredPlaylist.name,
    image: monitoredPlaylist.images[0],
  }));

export const unmonitoredPlaylists: Playlist[] = spotifyPlaylists
  .filter(
    (spotifyPlaylist) =>
      !databasePlaylists.some(
        (databasePlaylist) => databasePlaylist.id === spotifyPlaylist.id
      )
  )
  .map((monitoredPlaylist) => ({
    id: monitoredPlaylist.id,
    href: monitoredPlaylist.href,
    name: monitoredPlaylist.name,
    image: monitoredPlaylist.images[0],
  }));
