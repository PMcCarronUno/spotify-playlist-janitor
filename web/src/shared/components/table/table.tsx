import { TableBody } from "./table-body";
import { TableHead } from "./table-head";
import { useSortableTable } from "./hooks/use-sortable-table";
import { Column } from "./table-types";
import styled from "styled-components";

type TableProps = {
  caption: string;
  data: any[];
  columns: Column[];
};

const TableComponent = styled.table`
  border-collapse: collapse;
  width: 100%;

  th {
    border: none !important;
    border-bottom: 1px solid #121212 !important;
    min-width: 34px;
    text-align: left;
    color: white;
  }

  tbody tr:nth-child(even) {
    background-color: #121212;
  }

  tbody tr:hover {
    background-color: #232323;
  }

  td {
    border: none !important;
    padding: 8px;
  }
`;

export const Table = ({ caption, data, columns }: TableProps) => {
  const [tableData, handleSorting] = useSortableTable(data, columns);

  return (
    <TableComponent>
      {caption && <caption>{caption}</caption>}
      <TableHead {...{ columns, handleSorting }} />
      <TableBody {...{ columns, tableData }} />
    </TableComponent>
  );
};

Table.displayName = "Table";
