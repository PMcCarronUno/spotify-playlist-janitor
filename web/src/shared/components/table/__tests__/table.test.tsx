import { render } from "@testing-library/react";
import { tableData, tableColumns } from "shared/mock-data/table-data";

import { Table } from "../table";

describe("<Table />", () => {
  let container: HTMLElement;
  let props: any;

  beforeEach(() => {
    props = {
      caption: "Table caption.",
      data: tableData,
      columns: tableColumns,
    };
    ({ container } = render(<Table {...props} />));
  });

  it("should render Table component", () => {
    expect(container.firstChild).toMatchSnapshot();
  });
});
