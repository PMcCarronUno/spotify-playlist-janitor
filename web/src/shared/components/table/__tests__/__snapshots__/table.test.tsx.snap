// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should render Table component 1`] = `
<table
  class="sc-gLLuHO ehSqzN"
>
  <caption>
    Table caption.
  </caption>
  <thead>
    <tr
      class="sc-jTQDJr hyFECk"
    >
      <th
        class=""
        data-testid="table-head-cell-image"
        id="table-head-cell-image"
      >
        <div
          class="content"
        >
          <div
            class="sc-dmyDmy dcXriK"
          />
        </div>
      </th>
      <th
        class="sortable"
        data-testid="table-head-cell-title"
        id="table-head-cell-title"
      >
        <div
          class="content"
        >
          <div
            class="sc-dmyDmy dcXriK"
          >
            Title
          </div>
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M3.147 9l5 5h.707l5-5-.707-.707L9 12.439V2H8v10.44L3.854 8.292 3.147 9z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </th>
      <th
        class="sortable"
        data-testid="table-head-cell-artist"
        id="table-head-cell-artist"
      >
        <div
          class="content"
        >
          <div
            class="sc-dmyDmy dcXriK"
          >
            Artist
          </div>
        </div>
      </th>
      <th
        class="sortable"
        data-testid="table-head-cell-album"
        id="table-head-cell-album"
      >
        <div
          class="content"
        >
          <div
            class="sc-dmyDmy dcXriK"
          >
            Album
          </div>
        </div>
      </th>
      <th
        class="sortable"
        data-testid="table-head-cell-length"
        id="table-head-cell-length"
      >
        <div
          class="content"
        >
          <div
            class="sc-dmyDmy dcXriK"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 256 256"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M128,24A104,104,0,1,0,232,128,104.11,104.11,0,0,0,128,24Zm0,192a88,88,0,1,1,88-88A88.1,88.1,0,0,1,128,216Zm64-88a8,8,0,0,1-8,8H128a8,8,0,0,1-8-8V72a8,8,0,0,1,16,0v48h48A8,8,0,0,1,192,128Z"
              />
            </svg>
          </div>
        </div>
      </th>
      <th
        class="sortable"
        data-testid="table-head-cell-date_skipped"
        id="table-head-cell-date_skipped"
      >
        <div
          class="content"
        >
          <div
            class="sc-dmyDmy dcXriK"
          >
            Date Skipped
          </div>
        </div>
      </th>
      <th
        class=""
        data-testid="table-head-cell-delete"
        id="table-head-cell-delete"
      >
        <div
          class="content"
        >
          <div
            class="sc-dmyDmy dcXriK"
          />
        </div>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        data-testid="table-body-cell-0-image"
        id="table-body-cell-0-image"
      >
        <div
          class="sc-dmyDmy dcXriK"
        >
          <img
            alt=""
            height="40"
            src="https://i.scdn.co/image/ab67616d00004851176ca6e2a6efb0e3aaf4e37a"
            width="40"
          />
        </div>
      </td>
      <td
        data-testid="table-body-cell-0-title"
        id="table-body-cell-0-title"
      >
        <div
          class="sc-dmyDmy dcXriK"
        >
          Touch It / Technologic
        </div>
      </td>
      <td
        data-testid="table-body-cell-0-artist"
        id="table-body-cell-0-artist"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          Daft Punk
        </div>
      </td>
      <td
        data-testid="table-body-cell-0-album"
        id="table-body-cell-0-album"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          Alive 2007
        </div>
      </td>
      <td
        data-testid="table-body-cell-0-length"
        id="table-body-cell-0-length"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          4:37
        </div>
      </td>
      <td
        data-testid="table-body-cell-0-date_skipped"
        id="table-body-cell-0-date_skipped"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          Oct 15 2023 at 15:30:47
        </div>
      </td>
      <td
        data-testid="table-body-cell-0-delete"
        id="table-body-cell-0-delete"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            style="cursor: pointer;"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 3h3v1h-1v9l-1 1H4l-1-1V4H2V3h3V2a1 1 0 0 1 1-1h3a1 1 0 0 1 1 1v1zM9 2H6v1h3V2zM4 13h7V4H4v9zm2-8H5v7h1V5zm1 0h1v7H7V5zm2 0h1v7H9V5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </td>
    </tr>
    <tr>
      <td
        data-testid="table-body-cell-1-image"
        id="table-body-cell-1-image"
      >
        <div
          class="sc-dmyDmy dcXriK"
        >
          <img
            alt=""
            height="40"
            src="https://i.scdn.co/image/ab67616d00004851835a9e77dae1c928f871ac73"
            width="40"
          />
        </div>
      </td>
      <td
        data-testid="table-body-cell-1-title"
        id="table-body-cell-1-title"
      >
        <div
          class="sc-dmyDmy dcXriK"
        >
          Longshot
        </div>
      </td>
      <td
        data-testid="table-body-cell-1-artist"
        id="table-body-cell-1-artist"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          Catfish and the Bottlemen
        </div>
      </td>
      <td
        data-testid="table-body-cell-1-album"
        id="table-body-cell-1-album"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          The Balance
        </div>
      </td>
      <td
        data-testid="table-body-cell-1-length"
        id="table-body-cell-1-length"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          3:53
        </div>
      </td>
      <td
        data-testid="table-body-cell-1-date_skipped"
        id="table-body-cell-1-date_skipped"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          Oct 15 2023 at 15:30:45
        </div>
      </td>
      <td
        data-testid="table-body-cell-1-delete"
        id="table-body-cell-1-delete"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            style="cursor: pointer;"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 3h3v1h-1v9l-1 1H4l-1-1V4H2V3h3V2a1 1 0 0 1 1-1h3a1 1 0 0 1 1 1v1zM9 2H6v1h3V2zM4 13h7V4H4v9zm2-8H5v7h1V5zm1 0h1v7H7V5zm2 0h1v7H9V5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </td>
    </tr>
    <tr>
      <td
        data-testid="table-body-cell-2-image"
        id="table-body-cell-2-image"
      >
        <div
          class="sc-dmyDmy dcXriK"
        >
          <img
            alt=""
            height="40"
            src="https://i.scdn.co/image/ab67616d00004851d86a1e021e7acc7c07c7d668"
            width="40"
          />
        </div>
      </td>
      <td
        data-testid="table-body-cell-2-title"
        id="table-body-cell-2-title"
      >
        <div
          class="sc-dmyDmy dcXriK"
        >
          Live Forever
        </div>
      </td>
      <td
        data-testid="table-body-cell-2-artist"
        id="table-body-cell-2-artist"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          Oasis
        </div>
      </td>
      <td
        data-testid="table-body-cell-2-album"
        id="table-body-cell-2-album"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          Definitely Maybe
        </div>
      </td>
      <td
        data-testid="table-body-cell-2-length"
        id="table-body-cell-2-length"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          5:02
        </div>
      </td>
      <td
        data-testid="table-body-cell-2-date_skipped"
        id="table-body-cell-2-date_skipped"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          Oct 15 2023 at 15:30:47
        </div>
      </td>
      <td
        data-testid="table-body-cell-2-delete"
        id="table-body-cell-2-delete"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            style="cursor: pointer;"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 3h3v1h-1v9l-1 1H4l-1-1V4H2V3h3V2a1 1 0 0 1 1-1h3a1 1 0 0 1 1 1v1zM9 2H6v1h3V2zM4 13h7V4H4v9zm2-8H5v7h1V5zm1 0h1v7H7V5zm2 0h1v7H9V5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </td>
    </tr>
    <tr>
      <td
        data-testid="table-body-cell-3-image"
        id="table-body-cell-3-image"
      >
        <div
          class="sc-dmyDmy dcXriK"
        >
          <img
            alt=""
            height="40"
            src="https://i.scdn.co/image/ab67616d0000485158406b3f1ac3ceaff7a64fef"
            width="40"
          />
        </div>
      </td>
      <td
        data-testid="table-body-cell-3-title"
        id="table-body-cell-3-title"
      >
        <div
          class="sc-dmyDmy dcXriK"
        >
          Dark Necessities
        </div>
      </td>
      <td
        data-testid="table-body-cell-3-artist"
        id="table-body-cell-3-artist"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          Red Hot Chili Peppers
        </div>
      </td>
      <td
        data-testid="table-body-cell-3-album"
        id="table-body-cell-3-album"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          The Getaway
        </div>
      </td>
      <td
        data-testid="table-body-cell-3-length"
        id="table-body-cell-3-length"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          5:02
        </div>
      </td>
      <td
        data-testid="table-body-cell-3-date_skipped"
        id="table-body-cell-3-date_skipped"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          Oct 15 2023 at 15:30:46
        </div>
      </td>
      <td
        data-testid="table-body-cell-3-delete"
        id="table-body-cell-3-delete"
      >
        <div
          class="sc-hLQTFJ gjIzVE"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            style="cursor: pointer;"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M10 3h3v1h-1v9l-1 1H4l-1-1V4H2V3h3V2a1 1 0 0 1 1-1h3a1 1 0 0 1 1 1v1zM9 2H6v1h3V2zM4 13h7V4H4v9zm2-8H5v7h1V5zm1 0h1v7H7V5zm2 0h1v7H9V5z"
              fill-rule="evenodd"
            />
          </svg>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
